contributors/design-proposals/api-paging.md

```diff
@@ -0,0 +1,148 @@
+# Reduce the amount of client and server memory allocated during large API lists
+
+On large clusters, performing API queries that return all of the objects of a given resource type (GET /api/v1/pods, GET /api/v1/secrets) can lead to significant variations in peak memory use on the server and contribute substantially to long tail request latency.
+
+When loading very large sets of objects -- some clusters are now reaching 100k pods or equivalent numbers of supporting resources -- the system must:
+
+* Construct the full range description in etcd in memory and serialize it as protobuf in the client
+  * Some clusters have reported over 500MB being stored in a single object type
```

Comment:
>  I think leaving this to the client may solve only a part of the issue if the client decides to not send the limit query.

A client that does not send a limit query may be incapable of understanding partial responses with continue tokens. We cannot break compatibility with such clients.